### ISSUE TITLE
[Due for payment 2025-02-18] [Due for payment 2025-02-13] [Search v2.5] [Autocomplete] Highlight autocomplete key and selected value

### ISSUE THREAD

**Issue by lakchote** [Expensify/App#50949]
See https://docs.google.com/document/d/1o-Hp-tK8Z_BAcE-KRiXQicPH04qNr525EIxlG8J4RxM/edit?tab=t.0#bookmark=id.hy4h27dpoz37
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021864412596340849116</li>
        <li>Upwork Job ID: 1864412596340849116</li>
        <li>Last Price Increase: 2024-12-04</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sonialiap</details>

**Comment by 289Adam289** [comment]
Hi! According design doc I think this issue should be on hold for https://github.com/Expensify/react-native-live-markdown/pull/439 pr that will allow our parser to be passed as worklet to live markdown.

**Comment by ikevin127** [comment]
Not overdue, on hold for https://github.com/Expensify/react-native-live-markdown/pull/439.

**Comment by ikevin127** [comment]
Still on hold for https://github.com/Expensify/react-native-live-markdown/pull/439.

**Comment by ikevin127** [comment]
Still on hold for https://github.com/Expensify/react-native-live-markdown/pull/439.

**Comment by ikevin127** [comment]
Still on hold for https://github.com/Expensify/react-native-live-markdown/pull/439.

**Comment by lakchote** [comment]
> Still on hold for [Expensify/react-native-live-markdown#439](https://github.com/Expensify/react-native-live-markdown/pull/439).

Same.

**Comment by Kicu** [comment]
## Issue update

**Quick reminder:**
we wanted to use https://github.com/Expensify/react-native-live-markdown to implement highlighting in Search autocomplete. I recently discussed the best way to implement this with @tomekzaw who is the author of the PR.
Tomek has done a great deal of work 👏 to allow for the LiveMarkdown component to accept any JS code as parser via a prop.
Now we need to actually implement this in Expensify/App, which will require updating the version of live markdown (and some other packages) and tweaking the current code for `RNMarkdownTextInput` component.

This should not be a very big change in the code, but as any change it might introduce some small bugs.
In the old implementation of livemarkdown, the ExpensiMark parser was bundled together with the `MarkdownTextInput` component and was always used.
In the new version of `MarkdownTextInput`, the component accepts a `parser` prop, which can be any JS function adhering to correct interface.

I will try to push this forward as smoothly as possible, but we will require some help with testing.

Next steps :
1. I'm testing if the new version works correctly with Expensify locally (<--- we are here)
2. We will want to create a test build, and would need your help in getting some QAs to test it and see if there are no regressions for Composer with ExpensiMark.
3. merge https://github.com/Expensify/react-native-live-markdown/pull/439/ and release new version
4. bump version of `react-native-live-markdown` in E/App; pass `expensiMark` as parser for Composer, pass `autocompleteParser` for SearchRouter
5. Profit $$$ 😉 

We will also need to bump `reanimated` and `expensify-common` (minor version for both). Thankfully @blazejkustra is already looking at bumping reanimated here: https://github.com/Expensify/App/issues/52024 so this will speed things up.

I will open a draft PR soon for tracking the progress.
Feel free to ask anything. The workletization of `parser` in LiveMarkdown editor is a feature that multiple people would like to see added to the library. It will give us a lot of freedom, if we want to change how our parsers behave in future.

CC @luacmartins @lakchote @JmillsExpensify @tomekzaw @289Adam289 @SzymczakJ

**Comment by blazejkustra** [comment]
> We will also need to bump reanimated and expensify-common (minor version for both). Thankfully @blazejkustra is already looking at bumping reanimated here: https://github.com/Expensify/App/issues/52024 so this will speed things up.

I created the [issue](https://github.com/Expensify/App/issues/52024) earlier today, would you mind assigning me? @lakchote

**Comment by luacmartins** [comment]
Nice! Thanks for the detailed updated @Kicu! Looking forward to the PR and testing!

**Comment by lakchote** [comment]
Not overdue, progress still happening!

**Comment by lakchote** [comment]
> Not overdue, progress still happening!

Same

**Comment by luacmartins** [comment]
I appreciate the update @Kicu!

**Comment by lakchote** [comment]
[PR](https://github.com/software-mansion/react-native-reanimated/pull/6706) got merged, retest can be done.

**Comment by luacmartins** [comment]
Did we bump the version in App already?

**Comment by blazejkustra** [comment]
Not yet, [PR](https://github.com/Expensify/App/pull/52825) is still in progress. We'll update with more info tomorrow 👍

**Comment by blazejkustra** [comment]
Update: We need to wait for Reanimated 3.16.3, which includes a minor fix for shared value mocks that's necessary to proceed (tests are currently failing). It should be ready early next week, but if not, we can overwrite the mock ourselves = meaning it won't hold this task anyway.

**Comment by Kicu** [comment]
Meanwhile a few days ago an important PR for live-markdown was merged, that moved it from yarn to npm: https://github.com/Expensify/react-native-live-markdown/pull/539

There is a lot of things happening in live-markdown now, but the worklets PR is very close to done and we will be pushing it this week.

**Comment by lakchote** [comment]
Updates are just above.

**Comment by blazejkustra** [comment]
Update: [Reanimated bump](https://github.com/Expensify/App/pull/52825) is ready for review 🚀

**Comment by ikevin127** [comment]
@luacmartins Mind assigning BZ here for payment ? I'd appreciate it.

**Comment by luacmartins** [comment]
I assigned BZ. However, this issue is still not done since we gotta incorporate the changes in App

**Comment by Kicu** [comment]
The worklet PR got merged (https://github.com/Expensify/react-native-live-markdown/pull/439) 🚀 
Great work from @tomekzaw and @j-piasecki and also @289Adam289 who did a lot of testing.

Very soon someone will prepare a PR that bumps this uses this version with worklets in E/App, and then we can go forward from there.
We believe it will be safest to do two steps:
1. bump react-native-live-markdown and use it for existing Composer (the exact same markdown parser from expensify-common will be passed as prop instead of bundled together)
2. When we're sure there are no errors ☝️, then we use it in `SearchInput` and pass down our parser

**Comment by luacmartins** [comment]
That's great! I agree with the proposed approach. Hopefully we can get it working soon. I'll remove the hold on this issue since the live-markdown PR

**Comment by luacmartins** [comment]
>Does this mean BZ should hold payment for my review on the PR mentioned in https://github.com/Expensify/App/issues/50949#issuecomment-2506249175 or it shouldn't matter since I should be paid anyway ?
Or you mentioned that for the BZ to not close the issue after I'm paid since it's still ongoing ?

I think we can just process payment once when the issue is done. Not too set on this though

**Comment by Kicu** [comment]
chill out bot, I'm working on a PR today

**Comment by sonialiap** [comment]
Shhhh Melvin 😂

**Comment by Kicu** [comment]
It's fine, I can be assigned to this alongside others 👍

**Comment by sonialiap** [comment]
I'm OOO Dec 16-20, but I don't think this will need a BZ before I'm back so not adding a buddy

**Comment by luacmartins** [comment]
@Kicu https://github.com/Expensify/App/pull/53627 is merged. I think we can start working on this now

**Comment by Kicu** [comment]
Yup one of us will start working on this task this week. More updates will come soon

**Comment by 289Adam289** [comment]
Hi, I am Adam from SWM and I would like to work on this issue.

**Comment by luacmartins** [comment]
Nice, really excited for this one. Thanks for working on it @289Adam289!

**Comment by 289Adam289** [comment]
## Update:
I've been working on this issue and I'd like to share where we're at. Although is it WIP I’ve made significant progress. Below are some issues encountered and the advances made:

1. Worklets Compatibility: The parsers generated by Peggy use syntax incompatible with worklets (function.prototype), which complicates workletization. Thus, I wrote a script that modifies the autocomplete.js file.

2. Handling Spaces in LiveMarkdownTextInput: On the web, LiveMarkdownTextInput is implemented using a "contenteditable" div. LiveMarkdown simulates single-line input by using different css props e.g.`white-space: nowrap`  which collapses excess spaces and replaces them with `&nbsp;`. I have improved the parser to work with both types of spaces and changed the autocomplete values to include a trailing non-break space, which is not removed  by `white-space: nowrap`.

3. Selection Logic in LiveMarkdown: Currently, LiveMarkdown on the web lacks the functionality to adjust the selection dynamically as input length changes. I have implemented this feature, which was necessary only for the web version as the native implementation utilizes different native components.

4. Styling: I am currently focusing on improving the styling for web implementation. Markdown formatting is slightly moved in relation to the text, which is probably LM bug. @Kicu has already discussed it with LiveMarkdown team.

### Next Steps:

1. Improve styling on web.
2. Implement different markdown formatting depending on autocomplete key.

## Discussion Point:
It might be a good time to talk about how we format markdown for different types of keys. Right now, everything recognized by the autocomplete parsers gets the same blue highlight like "user-mention." Should we keep it this way, or try something different for various keys?
Looking forward to your thoughts and any further directions on how to proceed.
cc @luacmartins for thoughts

####  Videos to show how things look right now:

https://github.com/user-attachments/assets/24f9e59c-e2be-4c4c-af13-c23544a9ea99

https://github.com/user-attachments/assets/2fac4ec8-547c-4b2c-b2b0-5d4de27ad02c

https://github.com/user-attachments/assets/795ca935-e973-48ca-b784-effa6f3d09e0

**Comment by luacmartins** [comment]
@289Adam289 thanks for the update! 

>It might be a good time to talk about how we format markdown for different types of keys. Right now, everything recognized by the autocomplete parsers gets the same blue highlight like "user-mention." Should we keep it this way, or try something different for various keys?
Looking forward to your thoughts and any further directions on how to proceed.
cc @luacmartins for thoughts

The doc doesn't mention any specific styles, but the mocks seem to follow the `mentions` styles, i.e. everything has a blue highlight, except your own mention, which is green. So I think we can keep the same here and have everything blue, except for your own email/display name which should be green.

**Comment by luacmartins** [comment]
This is still a daily. SWM is coming back tomorrow and will resume work here.

**Comment by 289Adam289** [comment]
### Update:
Right now iOS and Android work well. Color of a highlight changes to green for current user mention. The only thing left are some cursor position and background alignment issues on web that are being worked on by @BartoszGrajdek in https://github.com/Expensify/react-native-live-markdown/pull/594.
I think we should wait for that LM pr. What do you think @luacmartins?

Current state:
Web:

https://github.com/user-attachments/assets/ebae6324-9b62-46f1-a93b-3b817e2195d2

Android:

https://github.com/user-attachments/assets/c26ac083-5606-4f28-a86a-761618555db6

iOS:

https://github.com/user-attachments/assets/e049d7d7-a5a9-4b54-b23e-334c8d82ce3b

**Comment by luacmartins** [comment]
Yea, I agree that we should wait for that PR to fix the alignment. Additionally, gonna tag @Expensify/design for more input on the highlight styles.

**Comment by dannymcclain** [comment]
I think the highlights are looking pretty good but it would be ideal if we could match them exactly to the styles we use for mention highlights (mainly thinking the border radius and subtle spacing).

Also this might be off topic so feel free to ignore if it is, but will the highlight always happen right away? I think I would expect it to only highlight if we found a legitimate match for the query—but I could be wrong about that. Will let the other designers gut check me there.

**Comment by dubielzyk-expensify** [comment]
Agree with both of Danny's comments. I also kinda expect it to only show highlight once it's matched, but I don't think it's suuuper critical.

**Comment by shawnborton** [comment]
Agree with both of those comments.

Also, when you are doing autocomplete on the second item, I would expect the full query to still be there. For example, this is doing two autocompletes in Slack:
![CleanShot 2025-01-09 at 10 01 50@2x](https://github.com/user-attachments/assets/37d6a12d-cfc6-4fbf-83e6-27071e45b1ae)

But this is what I see from the PR: 
![CleanShot 2025-01-09 at 10 02 32@2x](https://github.com/user-attachments/assets/89fb91cd-8740-47ef-bff7-695192534dc5)

**Comment by Kicu** [comment]
@shawnborton bot the top option to pick from is exactly the same query, just like on slack. Do you mean you would like for every single autocomplete item to contain the whole query?
Seems kinda weird to me. What if queries get super long?

**Comment by Kicu** [comment]
<img width="600" alt="Screenshot 2025-01-09 at 12 41 25" src="https://github.com/user-attachments/assets/5ec70f8d-711d-43a1-b208-87b3fc4660a9" />

It gets cut away on slack desktop app - not the best UX imo

**Comment by shawnborton** [comment]
Hmm yeah that's a fair point... though we could potentially right-align those long strings. 

Either way, I don't feel too strongly, let's see if others have any strong feelings otherwise we can just go with what you have planned already.

**Comment by dannymcclain** [comment]
I don't think I feel too strongly about that one 🤔 We could always go with the current plan and see how it feels first.

**Comment by dubielzyk-expensify** [comment]
Yeah, going with the current plan and adjusting feels right to me given it's hard to predict this one

**Comment by 289Adam289** [comment]
In response to the first @dannymcclain comment, the styles currently used by SearchRouter are the same as those used by Composer input. Adding border radius to the background is a bigger problem that is currently worked on in Expensify.

Regarding the issue of highlighting only when there's a legitimate query match, I think it would be safest to handle that as a follow-up. This PR is already complex, and adding further complications could increase the chance of missing regressions.

**Comment by Kicu** [comment]
RE: highlighting a mention only when it _actually_ mentions a real login.
This is not something that we are able to easily do in the app right now, and it's a bigger task that Im currently working on in the context of Composer (the main chat input). Here is the issue: https://github.com/Expensify/App/issues/38025

We will make this work in Search input once the above task is completed. We will also make sure that the logic and look of highlights in Search and Composer is the same.

**Comment by dannymcclain** [comment]
Thanks for the explanation—I'm fine handling that stuff as a follow up since it sounds like this PR is already pretty intense.

**Comment by shawnborton** [comment]
Yup, that totally works for me too - thanks for clarifying things.

**Comment by Kicu** [comment]
This is being actively worked on. While Adam focused on most of this task and logic of working, I need to do a bit more work related to LiveMarkdown input to fix the styling problems. We also have colleagues from live markdown helping with this.

**Comment by luacmartins** [comment]
^ what he said

**Comment by 289Adam289** [comment]
All issues with LiveMarkdown's single-line on the web are resolved. `react-native-live-markdown` package is being updated to a new version that includes fix in https://github.com/Expensify/App/pull/55509

**Comment by luacmartins** [comment]
Nice! Hopefully we can get that PR merged soon and continue work here.

**Comment by Kicu** [comment]
We are also waiting for this one: https://github.com/Expensify/App/pull/55343 but I hope it can be merged soon.
Without it the text won't be aligned vertically.

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/pull/55343 is merged! Let's keep up the momentum!

**Comment by Kicu** [comment]
I provided fix for the styling regression after my PR, so we don't have to revert anything.
Currently waiting only on LiveMarkdown bump which fixes some more styling.

We are 90% there :)

**Comment by luacmartins** [comment]
Nice! Thanks for working on a fix. Is this the live markdown bump PR we're waiting on? https://github.com/Expensify/App/pull/55706

**Comment by luacmartins** [comment]
Oh I see, we also have [this one](https://github.com/Expensify/App/pull/55509)

**Comment by Kicu** [comment]
Correct it's this one ☝

**Comment by ikevin127** [comment]
No payment for me here, instead @rayane-djouah was C+ reviewer for the PR above.

**Comment by 289Adam289** [comment]
@luacmartins the LiveMarkdown bump has been merged!  I've opened the pr for review

**Comment by luacmartins** [comment]
Amazing! I pinged C+ and I'll also review it today.

**Comment by luacmartins** [comment]
Just for posteriority, we need to address the following features after [this PR](https://github.com/Expensify/App/pull/54403) is merged:

1. The syntax keys, e.g. in, from, type, should be in supporting text color
2. Only highlight exact autocomplete values

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/pull/54403 is merged! Great work @289Adam289! We can start working on the follow ups: https://github.com/Expensify/App/issues/50949#issuecomment-2622068254.

**Comment by 289Adam289** [comment]
@luacmartins  I've prepared a draft PR that addresses both follow-up issues. There's one more thing that needs to be addressed before I can open it: one more LiveMarkdown bump to version 0.219, which includes fixes to issues I've encountered during work on the follow-ups. Unfortunately, the workflow publishing the package to npm has failed, probably because of an expired token.

I will open a PR with a bump as soon as the package is available on npm. Meanwhile, here is a video showing how everything looks with the package installed manually.

https://github.com/user-attachments/assets/a117e4a3-08ed-4034-b7ae-ec64b761d83a

Let me know what you think?

**Comment by Kicu** [comment]
I also have to say this looks super neat, and also quite fast because of great work of our friends from Live markdown 😁 
CC @tomekzaw @BartoszGrajdek

**Comment by luacmartins** [comment]
Work is ongoing on this PR https://github.com/Expensify/App/pull/56243

**Comment by ikevin127** [comment]
🔄 Payment for the first PR I reviewed ([here](https://github.com/Expensify/App/issues/50949#issuecomment-2641149572)) is **due today 13th Feb** and for the second one is due on the 18th Feb ([here](https://github.com/Expensify/App/issues/50949#issuecomment-2649659074)).

cc @sonialiap for visibility

**Comment by mallenexpensify** [comment]
@ikevin127 can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~021890456480441028724

**Comment by ikevin127** [comment]
@mallenexpensify Offer accepted, thank you! ❤️

**Comment by mallenexpensify** [comment]
Upwork is bugging out right now, will try to get back to to pay later today.

**Comment by mallenexpensify** [comment]
Contributor+: @ikevin127 paid $250 via Upwork
